### ISSUE TITLE
Investigate native mode for the operator

### ISSUE THREAD

**Issue by vmuzikar** [keycloak/keycloak#9436]
To resolve this issue the following questions need to be answered.
* What is the perf of native mode in comparison to JVM mode in case of a lightweight app like an operator?
* Are there any blockers for using native mode?
* How does it align with support for IBM platforms?

**Comment by vmuzikar** [comment]
@andreaTP @jonathanvila @pedroigor @DGuhr jFYI

**Comment by manusa** [comment]
> That said, JVM mode is probably more suited for operators because of their typically long-running lifecycle similar to typical Java application servers for which the JVM is particularly optimized

^^^ This

> do things seem relatively stable now?

Stability-wise, I'd assume things are mostly working now. It's been a while since I've seen any native-related issue reported either on the Fabric8 or Quarkus repository.

> And I've reduced image sizes from ~300MB to ~50MB and startup times from ~15s to ~0.1s

In case you want to cater for those users too, then you might just want to include two options for the operator: native and JVM.

You should then document the advantages and recommended use of each alternative.
If the user requires performance and high throughput at the expense of cloud costs, then use JVM.
If the user is concerned about cloud costs and doesn't care that much about performance (the  operator is handling very few installations), then use native.

**Comment by heruan** [comment]
> Operators usually don't require to be restarted often (so starting time or image size aren't usually an issue).

It's not just restarts, consider for example the following scenario: when deploying an app that needs Keycloak, CloudNativePG and CertManager on a new namespace, you need first to deploy the operator and then the Keycloak CR that are both provided as JVM images. If the app needs to connect to Keycloak at startup, the deployment workflow has to wait for the Keycloak CR to be ready, that in turn needs for the Keycloak Operator to be ready (being scoped to the namespace).

This is a comparison of image sizes and startup times of some of the operators we use, where both Keycloak Operator and CR images are causing some startup delay:

| Image                              | Size (MB) | Startup time (s) |
|:-----------------------------------|----------:|-----------------:|
| `jetstack/cert-manager-controller` | 70        | 0.01             |
| `cloudnative-pg/cloudnative-pg`    | 117       | 0.3              |
| `keycloak/keycloak-operator`       | 426       | 7.427            |
| `keycloak/keycloak`                | 444       | 13.979           |

If going native has some side-effects (I'm not aware of that), it would be great to improve images size and startup times of both the Operator and CR images.

**Comment by metacosm** [comment]
Makes sense, such a deployment topology is indeed problematic. However, an operator could also be deployed once for the cluster and be able to deploy resources on specified namespaces without needing to be deployed once per namespace…

**Comment by heruan** [comment]
> Makes sense, such a deployment topology is indeed problematic. However, an operator could also be deployed once for the cluster and be able to deploy resources on specified namespaces without needing to be deployed once per namespace…

Indeed, but this isn't the case for the Keycloak Operator… right? Or at least by default it does watch over namespaces rather than its own.

**Comment by shawkins** [comment]
> Indeed, but this isn't the case for the Keycloak Operator… right? Or at least by default it does not watch over namespaces other than its own.

Correct, we don't currently promote the ability to watch all namespaces with the keycloak operator.
